pulsar-broker/src/main/java/org/apache/pulsar/utils/SimpleCache.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class SimpleCache<K, V> {
```

Comment:
```
public class SimpleCache<K, V> {
    private class CacheEntry {
        V value;
        long lastAccessTime;
        Consumer<V> expireCallback;

        CacheEntry(V value, Consumer<V> expireCallback) {
            this.value = value;
            this.expireCallback = expireCallback;
            this.lastAccessTime = System.currentTimeMillis();
        }
    }

    private final HashMap<K, CacheEntry> cache = new HashMap<>();
    private final ScheduledFuture<?> scheduledFuture;
    private final long evictionDelay;

    public SimpleCache(ScheduledExecutorService scheduler, long evictionDelay) {
        this.evictionDelay = evictionDelay;
        this.scheduledFuture =
                scheduler.scheduleAtFixedRate(this::evictionTask, 100, 100, TimeUnit.MILLISECONDS);
    }

    public synchronized V get(final K key, final Supplier<V> valueSupplier, final Consumer<V> expireCallback) {
        CacheEntry existingValue = cache.get(key);
        if (existingValue != null) {
            existingValue.lastAccessTime = System.currentTimeMillis();
            return existingValue.value;
        }
        final V value = valueSupplier.get();
        cache.put(key, new CacheEntry(value, expireCallback));
        return value;
    }

    private void evictionTask() {
        long now = System.currentTimeMillis();
        for (K key : cache.keySet()) {
            CacheEntry entry = cache.get(key);
            if (entry != null && (now - entry.lastAccessTime) > evictionDelay) {
                CacheEntry remove = cache.remove(key);
                if (remove != null) {
                    remove.expireCallback.accept(remove.value);
                }
            }
        }
    }

    public void shutdown() {
        scheduledFuture.cancel(true);
    }
}

```

Maybe we can use a `scheduleAtFixedRate` to check the key expiry

Reply:
I changed the implementation now. The main difference from your example here is that I didn't add a `shutdown` method because the shutdown of the executor will cancel all pending scheduled tasks.